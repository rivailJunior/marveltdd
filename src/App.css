.labelHeader {
  font: normal normal bold 32px/30px PT Sans Caption;
  letter-spacing: 0px;
  color: #555555;
  opacity: 1;
}