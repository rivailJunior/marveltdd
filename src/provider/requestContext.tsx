import { createContext, useContext, useMemo } from "react";

import {
    getParticipation,
} from "./service";

type RequestContextType = {
    getParticipation(id: number, type: string): Promise<any>;
};

const RequestContext = createContext({
    getParticipation,
});

export const useRequestContext = () => {
    const context = (useContext(
        RequestContext
    ) as unknown) as RequestContextType;
    return context;
};

export const RequestProvider = (props: any) => {
    const { getParticipation } = useRequestContext();
    const value = useMemo(() => {
        return { getParticipation }
    }, [])
    return (
        <RequestContext.Provider
            value={value}
            {...props}
        />
    );
};

export default RequestContext;
