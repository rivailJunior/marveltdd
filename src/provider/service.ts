import { Character, CharacterDataContainer } from './../characterTypes/characters';
import axios, { AxiosResponse } from 'axios'
const apikey = process.env.REACT_APP_PUBKEY

export enum ParticipationTypes {
    series = "series",
    events = "events",
    comics = "comics",
}

const api = axios.create({
    baseURL: 'https://gateway.marvel.com/v1/public',
});

const getService = async (url: string, options: object): Promise<AxiosResponse> => {
    return await api.get(url, { params: { apikey, ...options } })
}

const getCharacters = async (total: number, offset?: number): Promise<CharacterDataContainer> => {
    const currentPage = offset || 0;
    const result = await getService('/characters', { offset: currentPage, limit: total })
    return result.data?.data
}

const getParticipation = async (id: number, type: string): Promise<any> => {
    const result = await getService(`/characters/${ id }/${ type }`, {})
    return result.data?.data
}

const getCharacter = async (id: number): Promise<[Character]> => {
    const result = await getService(`/characters/${ id }`, {})
    return result.data?.data?.results
}

const getCharacterByName = async (name: string): Promise<CharacterDataContainer> => {
    const result = await getService(`/characters?nameStartsWith=${ name }`, {})
    return result.data?.data
}

export { getCharacters, getCharacter, getParticipation, getCharacterByName };