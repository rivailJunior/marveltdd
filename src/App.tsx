import "./App.css";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import Home from "./pages/home";
import Description from "./pages/description";
import { RequestProvider } from "./provider/requestContext";
import "bootstrap/dist/css/bootstrap.min.css";
import { MarvelHeader } from "./components/header/marvelHeader";
import { Provider } from 'react-redux'
import { store } from "./model/storeConfig";


function App(): JSX.Element {
    return (
        <Router>
            <Provider store={store}>
                <RequestProvider>
                    <MarvelHeader
                        leftLabel="Objective"
                        userName="Rivail  "
                        descriptionRight="Teste front-end"
                    />
                    <Switch>
                        <Route exact path="/">
                            <Home />
                        </Route>
                        <Route path="/description">
                            <Description />
                        </Route>
                        <Route path="/:id">
                            <Home />
                        </Route>
                    </Switch>
                </RequestProvider>
            </Provider>
        </Router>
    );
}

export default App;
