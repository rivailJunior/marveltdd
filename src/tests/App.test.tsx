import React from "react";
import {
    render,
    cleanup,
    // fireEvent,
    act,
    // screen,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import App from "../pages/home";
import { RequestProvider } from "../provider/requestContext";
import { getCharacters } from "../provider/service";

afterEach(cleanup);
// beforeEach(cleanup);

describe("`<App>`", () => {
    test("Render correctly", () => {
        const container = renderer.create(<App />);
        container.toJSON();
        expect(container).toMatchSnapshot();
    });

    test("Show information before loading list and after that show 10", async () => {
        const promise = getCharacters(10);
        const { getByText, findAllByTestId } = render(
            <RequestProvider values={{ getCharacters: promise }}>
                <App />
            </RequestProvider>
        );
        expect(getByText(/nenhum personagem encontrado/i)).toBeInTheDocument();

        await act(() => promise);
        // const listLi = await findAllByTestId("marvelLi");
        // expect(listLi).toHaveLength(10); // sometimes its return only one
    });
});
