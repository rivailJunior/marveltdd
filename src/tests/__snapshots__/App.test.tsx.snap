// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<App>\` Render correctly 1`] = `
<div>
  <div
    className="jumbotron"
  >
    <div
      className="container"
    >
      <h1
        className="labelHeader"
        data-testid="labelheader"
      >
        Busca de personagens
      </h1>
      <div
        className="row"
      >
        <div
          className="col-md-3"
        >
          <input
            data-testid="inputSearch"
            onChange={[Function]}
            type="text"
            value="Capitao america"
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-12 col-12"
        >
          <div
            className="labelHeader showList"
          >
            Nenhum personagem encontrado
          </div>
          <div
            className="hideList list-group list-group-flush"
            onKeyDown={[Function]}
          >
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "backgroundColor": "transparent",
                }
              }
            >
              <div
                className="row"
              >
                <div
                  className="col-md-2"
                />
                <div
                  className="col-md-2"
                >
                  <div
                    className="defaultLabel"
                  >
                    Personagem
                  </div>
                </div>
                <div
                  className="d-none d-lg-block col-md-4"
                >
                  <div
                    className="defaultLabel"
                  >
                    Séries
                  </div>
                </div>
                <div
                  className="d-none d-lg-block col-md-4"
                >
                  <div
                    className="defaultLabel"
                  >
                    Eventos
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="listGroup list-group"
            onKeyDown={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="paginationDiv"
  >
    <ul
      className="noPagination pagination"
      data-testid="paginationDiv"
    >
      <li
        className="paginationItem page-item"
      >
        <a
          className="page-link"
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <span
            aria-hidden="true"
          >
            ›
          </span>
          <span
            className="sr-only"
          >
            Next
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
